src/smt/proof_post_processor.h

```diff
@@ -149,6 +149,80 @@ class ProofPostprocessCallback : public ProofNodeUpdaterCallback
   bool addToTransChildren(Node eq,
                           std::vector<Node>& tchildren,
                           bool isSymm = false);
+
+  /**
+   * When given children and args lead to different sets of literals in a
+   * conclusion depending on whether macro resolution or chain resolution is
+   * applied, the literals in the difference of the second from the first, from
+   * now on "crowding literals", are literals removed implicitly by macro
+   * resolution. For example
+   *
+   *      l0 v l0 v l0 v l1 v l2    ~l0 v l1   ~l1
+   * (1)  ----------------------------------------- MACRO_RES
+   *                 l2
+   *
+   * but
+   *
+   *      l0 v l0 v l0 v l1 v l2    ~l0 v l1   ~l1
+   * (2)  ---------------------------------------- CHAIN_RES
+   *                l0 v l0 v l1 v l2
+   *
+   * where l0 and l1 are crowding literals in the second proof.
+   *
+   * There are two views for how MACRO_RES implicitly removes the crowding
+   * literal, i.e., how MACRO_RES can be expanded into CHAIN_RES so that
+   * crowding literals are removed. The first is that (1) becomes
+   *
+   *  l0 v l0 v l0 v l1 v l2  ~l0 v l1  ~l0 v l1  ~l0 v l1  ~l1  ~l1  ~l1  ~l1
+   *  ---------------------------------------------------------------- CHAIN_RES
+   *                                 l2
+   *
+   * via the repetition of the premise responsible for removing more than one
+   * occurrence of the crowding literal. The issue however is that this
+   * expansion is exponential. Note that (2) has two occurrences of l0 and one
+   * of l1 as crowding literals. However, by repeating ~l0 v l1 two times to
+   * remove l0, the clause ~l1, which would originally need to be repeated only
+   * one time, now has to be repeated two extra times on top of that one. With
+   * multiple crowding literals and their elimination depending on premises that
+   * themselves add crowding literals one can easily end up with resolution
+   * chains going from dozens to thousands of premises. Such examples do occur
+   * in practice, even in our regressions.
+   *
+   * The second way of expanding MACRO_RES, which avoids this exponential
+   * behavior, is so that (1) becomes
+   *
+   *      l0 v l0 v l0 v l1 v l2
+   * (4)  ---------------------- FACTORING
+   *      l0 v l1 v l2                       ~l0 v l1
+   *      ------------------------------------------- CHAIN_RES
```

Comment:
Since in general we may have more than two premises I always used chain resolution. I'm gonna add a note in the documentation, but I'd rather keep the code as is.